### PR TITLE
CLN refactor maybe-castable

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

extracted from #37586

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -1301,20 +1301,18 @@ def convert_dtypes(
     return inferred_dtype
 
 
-def maybe_castable(arr: np.ndarray) -> bool:
+def maybe_castable(dtype: DtypeObj) -> bool:
```
Comment (jbrockmendel) [inline pandas/core/dtypes/cast.py:1304]: i think just np.dtype?

Comment (MarcoGorelli) [reply pandas/core/dtypes/cast.py:1304]: in light of https://github.com/pandas-dev/pandas/pull/39104#discussion_r555062453 , maybe it's worth mothballing this too until we have numpy types

Comment (jreback) [reply pandas/core/dtypes/cast.py:1304]: kk let's revert this part

Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py:1304]: i dont understand why "just use `np.dtype`" isnt the right thing to do here

Comment (MarcoGorelli) [reply pandas/core/dtypes/cast.py:1304]: I thought the objection was that it resolves to Any:
```
$ cat t.py 
import numpy as np

a: np.dtype
reveal_type(a)
$ mypy t.py 
t.py:4: note: Revealed type is 'Any'
```

- is that OK?

Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py:1304]: wont that get fixed on numpy's end at some point?  i mean, np.dtype is the correct type here

**Comment by jreback** [comment]
thanks @MarcoGorelli

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/cast.py]: the problem (in general) is that we are adding a type that resolves to Any, so if there are mypy issues 'lurking' they will not be reported.

when we do have numpy types, we may then have more errors to resolve.

That train left the station long ago. so potentially having a few more errors is nbd.

this is a different scenario to https://github.com/pandas-dev/pandas/pull/39104#discussion_r555062453. That PR is focused on removing type ignores. In that case, the correct type is np.ufunc and not Callable, so again changing that is the right thing to do. just maybe not at the right time.

**Pull request merged**